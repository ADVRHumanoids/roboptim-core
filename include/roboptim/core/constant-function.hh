// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_CONSTANT_FUNCTION_HH
# define ROBOPTIM_CORE_CONSTANT_FUNCTION_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <roboptim/core/linear-function.hh>

namespace roboptim
{
  /// \addtogroup roboptim_function
  /// @{

  /// \brief Constant function.
  ///
  /// Implement a constant function using the formula:
  /// \f[f(x) = offset\f]
  /// where \f$offset\f$ is set when the class is instantiated.
  class ROBOPTIM_DLLAPI ConstantFunction : public LinearFunction
  {
  public:
    /// \brief Build an constant function.
    ///
    /// \param offset constant function offset
    ConstantFunction (const vector_t& offset) throw ();
    ~ConstantFunction () throw ();

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    void impl_compute (result_t& , const vector_t&) const throw ();
    void impl_gradient (gradient_t&, const vector_t&, size_type = 0)
      const throw ();
    void impl_jacobian (jacobian_t&, const vector_t&) const throw ();

  private:
    const vector_t offset_;
  };

  /// Example shows constant function use.
  /// \example constant-function.cc

  /// @}

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_CONSTANT_FUNCTION_HH
