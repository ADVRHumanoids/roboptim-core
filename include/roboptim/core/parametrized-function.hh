// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_PARAMETRIZED_FUNCTION_HH
# define ROBOPTIM_CORE_PARAMETRIZED_FUNCTION_HH
# include <iostream>
# include <limits>
# include <utility>

# include <boost/numeric/ublas/matrix.hpp>
# include <boost/numeric/ublas/vector.hpp>
# include <boost/tuple/tuple.hpp>

# include <roboptim/core/fwd.hh>

namespace roboptim
{
  namespace ublas = boost::numeric::ublas;

  /// \addtogroup roboptim_meta_function
  /// @{

  /// \brief Define an abstract parametrized mathematical function (\f$C^0\f$).
  ///
  /// A function is an object that can be evaluated for a given
  /// point.
  ///
  /// \f[ f : p \rightarrow g \f]
  /// \f$p \in \mathbb{R}^p\f$, \f$g \in G\f$ where
  /// \f$p\f$ is the input size and \f$G = \mathbb{R}^n \in \mathbb{R}^m\f$
  /// (\f$n, m \in \mathbb{R}^2\f$).
  ///
  /// Functions are pure immutable objects: evaluating a function
  /// twice at a given point <b>must</b> give the same result.
  template <typename F>
  class ParametrizedFunction
  {
  public:
    /// \brief Import value type.
    typedef typename F::value_type value_type;
    /// \brief Import size type.
    typedef typename F::size_type size_type;
    /// \brief Import vector type.
    typedef typename F::vector_t vector_t;
    /// \brief Import matrix type.
    typedef typename F::matrix_t matrix_t;
    /// \brief Import  result type.
    typedef F result_t;
    /// \brief Import argument type.
    typedef typename F::argument_t argument_t;

    /// \brief Evaluate the function at a specified point.
    ///
    /// The program will abort if the argument does not have the
    /// expected size.
    /// \param argument point at which the function will be evaluated
    /// \return computed result
    result_t operator () (const argument_t& argument) const throw ();

    /// \brief Return the input size (i.e. argument's vector size).
    ///
    /// \return input size
    size_type inputSize () const throw ();

    /// \brief Display the function on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();

  protected:
    /// \brief Concrete class constructor should call this constructor.
    ///
    /// \param inputSize function arity
    ParametrizedFunction (size_type inputSize) throw ();

    /// \brief Function evaluation.
    ///
    /// Evaluate the function, has to be implemented in concrete classes.
    /// \warning Do not call this function directly, call #operator() instead.
    /// \param argument point at which the function will be evaluated
    virtual result_t impl_compute (const argument_t& argument)
      const throw () = 0;

  private:
    /// Problem dimension.
    const size_type inputSize_;
  };
  /// @}


  /// \brief Override operator<< to handle function display.
  ///
  /// \param o output stream used for display
  /// \param f function to be displayed
  /// \return output stream
  template <typename F>
  std::ostream& operator<< (std::ostream& o, const ParametrizedFunction<F>& f);


} // end of namespace roboptim

# include <roboptim/core/parametrized-function.hxx>
#endif //! ROBOPTIM_CORE_PARAMETRIZED_FUNCTION_HH
