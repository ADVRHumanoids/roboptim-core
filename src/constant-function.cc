// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "debug.hh"

#include <boost/numeric/ublas/io.hpp>
#include <boost/numeric/ublas/matrix.hpp>
#include <boost/numeric/ublas/vector.hpp>

#include <roboptim/core/indent.hh>
#include <roboptim/core/constant-function.hh>

namespace roboptim
{
  ConstantFunction::ConstantFunction (const vector_t& offset)
    throw ()
    : LinearFunction (offset.size (), offset.size (), "constant function"),
      offset_ (offset)
  {
  }

  ConstantFunction::~ConstantFunction () throw ()
  {
  }

  void
  ConstantFunction::impl_compute (result_t& result,
				  const argument_t&)
    const throw ()
  {
    result = this->offset_;
  }

  void
  ConstantFunction::impl_jacobian (jacobian_t& jacobian,
				   const argument_t&) const throw ()
  {
    jacobian.clear ();
  }

  void
  ConstantFunction::impl_gradient (gradient_t& gradient,
				   const argument_t&,
				   size_type) const throw ()
  {
    gradient.clear ();
  }

  std::ostream&
  ConstantFunction::print (std::ostream& o) const throw ()
  {
    return o << "Constant function" << incindent << iendl
             << "offset = " << this->offset_ << iendl
             << decindent;
  }

} // end of namespace roboptim
