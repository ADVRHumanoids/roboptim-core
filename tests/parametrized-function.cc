// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>

#include "common.hh"

#include <roboptim/core/io.hh>
#include <roboptim/core/constant-function.hh>
#include <roboptim/core/parametrized-function.hh>

using namespace roboptim;

// Define a simple function.
struct ParametrizedF : public ParametrizedFunction<ConstantFunction>
{
  ParametrizedF () : ParametrizedFunction<ConstantFunction> (1, 1, 1)
  {}

  result_t impl_compute (const argument_t& argument) const throw ()
  {
    return result_t (argument);
  }
};

#define CHECKME(PVALUE)						\
  {								\
    std::cout << "Parameter is " << PVALUE << std::endl;	\
    parameter[0] = (PVALUE);					\
    ConstantFunction cst = pf (parameter);			\
								\
    ConstantFunction::vector_t x (1);				\
    x[0] = 31.;							\
    std::cout							\
      << cst << std::endl					\
      << "Evaluate: " << std::endl				\
      << cst (x) << std::endl					\
      << "Gradient: " << std::endl				\
      << cst.gradient (x) << std::endl				\
      << "Jacobian: " << std::endl				\
      << cst.jacobian (x) << std::endl				\
      << std::endl;						\
  }								\

int run_test ()
{
  ParametrizedF pf;

  ParametrizedF::argument_t parameter (1);
  parameter.clear ();

  CHECKME (0.);
  CHECKME (3.14);
  CHECKME (42.);

  {
    parameter[0] = 128.;
    ConstantFunction::vector_t x (1);
    x[0] = 256.;

    // Natural evaluation in one line.
    ConstantFunction::result_t res = pf (parameter) (x);

    assert (res[0] == 128.);
  }

  return 0;
}

GENERATE_TEST ()
