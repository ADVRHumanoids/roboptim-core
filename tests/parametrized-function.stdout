Parameter is 0
Constant function
  offset = [1](0)
  
Evaluate: 
[1](0)
Gradient: 
[1](0)
Jacobian: 
[1](0)

Parameter is 3.14
Constant function
  offset = [1](3.14)
  
Evaluate: 
[1](3.14)
Gradient: 
[1](0)
Jacobian: 
[1](0)

Parameter is 42
Constant function
  offset = [1](42)
  
Evaluate: 
[1](42)
Gradient: 
[1](0)
Jacobian: 
[1](0)

